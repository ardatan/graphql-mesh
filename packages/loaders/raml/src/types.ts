import { JSONSchemaLoaderOptions } from '@omnigraph/json-schema';

export interface RAMLLoaderOptions extends Partial<JSONSchemaLoaderOptions> {
  ramlFilePath: string;
}
