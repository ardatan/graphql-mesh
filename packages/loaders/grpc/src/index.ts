import type { Logger, YamlConfig } from "@graphql-mesh/types";
import { gRPCLoader } from "./gRPCLoader";

interface LoaderContext {
  cwd: string;
  logger: Logger;
}

export function loadGrpcSubgraph(name: string, options: YamlConfig.GrpcHandler) {
  return (ctx: LoaderContext) => ({
    name,
    schema$: new gRPCLoader(name, ctx.cwd, ctx.logger, options).buildSchema(),
  });
}
