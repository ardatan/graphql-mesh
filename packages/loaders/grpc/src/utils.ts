import { existsSync } from "fs";
import type { SchemaComposer } from "graphql-compose";
import { isAbsolute, join } from "path";
import type { INamespace, Root } from "protobufjs";
import { getGraphQLScalarForGrpc, isGrpcScalar } from "./scalars.js";
import lodashHas from 'lodash.has';

export function addIncludePathResolver(root: Root, includePaths: string[]): void {
  const originalResolvePath = root.resolvePath;
  root.resolvePath = (origin: string, target: string) => {
    if (isAbsolute(target)) {
      return target;
    }
    for (const directory of includePaths) {
      const fullPath: string = join(directory, target);
      if (existsSync(fullPath)) {
        return fullPath;
      }
    }
    const path = originalResolvePath(origin, target);
    if (path === null) {
      console.warn(`${target} not found in any of the include paths ${includePaths}`);
    }
    return path;
  };
}

export function getTypeName(
  schemaComposer: SchemaComposer,
  pathWithName: string[] | undefined,
  isInput: boolean,
) {
  if (pathWithName?.length) {
    const baseTypeName = pathWithName.filter(Boolean).join('__');
    if (isGrpcScalar(baseTypeName)) {
      return getGraphQLScalarForGrpc(baseTypeName);
    }
    if (schemaComposer.isEnumType(baseTypeName)) {
      return baseTypeName;
    }
    return isInput ? baseTypeName + '_Input' : baseTypeName;
  }
  return 'Void';
}

export function walkToFindTypePath(
  rootJson: INamespace,
  pathWithName: string[],
  baseTypePath: string[],
) {
  const currentWalkingPath = [...pathWithName];
  while (!lodashHas(rootJson.nested, currentWalkingPath.concat(baseTypePath).join('.nested.'))) {
    if (!currentWalkingPath.length) {
      break;
    }
    currentWalkingPath.pop();
  }
  return currentWalkingPath.concat(baseTypePath);
}
