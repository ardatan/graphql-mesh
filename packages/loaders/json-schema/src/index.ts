import { loadGraphQLSchemaFromJSONSchemas } from './loadGraphQLSchemaFromJSONSchemas.js';
export default loadGraphQLSchemaFromJSONSchemas;
export * from './loadGraphQLSchemaFromJSONSchemas.js';
export * from './getComposerFromJSONSchema.js';
export * from './getDereferencedJSONSchemaFromOperations.js';
export * from './getGraphQLSchemaFromDereferencedJSONSchema.js';
export * from './types.js';
export { processDirectives } from './directives.js';
export * from './bundle.js';
