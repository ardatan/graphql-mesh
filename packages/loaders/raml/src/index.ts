export { loadGraphQLSchemaFromRAML as default } from './loadGraphQLSchemaFromRAML';
export { getJSONSchemaOptionsFromRAMLOptions } from './getJSONSchemaOptionsFromRAMLOptions';
export { RAMLLoaderOptions } from './types';
export * from './bundle';
