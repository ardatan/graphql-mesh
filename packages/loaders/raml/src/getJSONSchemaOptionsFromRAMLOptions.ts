import { sanitizeNameForGraphQL } from '@graphql-mesh/utils';
import { HTTPMethod, JSONSchemaOperationConfig } from '@omnigraph/json-schema';
import { getAbsolutePath, getCwd, JSONSchemaObject } from 'json-machete';
import { api10, loadApi } from '@ardatan/raml-1-parser';
import { fetch as crossUndiciFetch } from 'cross-undici-fetch';
import toJsonSchema from 'to-json-schema';
import { RAMLLoaderOptions } from './types';

export async function getJSONSchemaOptionsFromRAMLOptions({
  ramlFilePath,
  cwd: ramlFileCwd = process.cwd(),
  operations: extraOperations,
  baseUrl: forcedBaseUrl,
  fetch = crossUndiciFetch,
  schemaHeaders = {},
}: RAMLLoaderOptions): Promise<{
  operations: JSONSchemaOperationConfig[];
  cwd: string;
  baseUrl: string;
  fetch?: WindowOrWorkerGlobalScope['fetch'];
}> {
  const operations = extraOperations || [];
  const ramlAbsolutePath = getAbsolutePath(ramlFilePath, ramlFileCwd);
  const ramlAPI = (await loadApi(ramlAbsolutePath, [], {
    httpResolver: {
      getResourceAsync: async (url: string) => {
        const fetchResponse = await fetch(url, {
          headers: schemaHeaders,
        });
        const content = await fetchResponse.text();
        return {
          content,
        };
      },
      getResource: () => {
        throw new Error(`Sync fetching not available for URLs`);
      },
    },
  })) as api10.Api;
  let baseUrl = forcedBaseUrl;
  if (!baseUrl) {
    baseUrl = ramlAPI.baseUri().value();
    for (const baseUrlParamNode of ramlAPI.baseUriParameters()) {
      const paramName = baseUrlParamNode.name();
      baseUrl = baseUrl.split(`{${paramName}}`).join(`{context.${paramName}}`);
    }
  }
  const pathTypeMap = new Map<string, string>();
  for (const typeNode of ramlAPI.types()) {
    const typeNodeJson = typeNode.toJSON();
    for (const typeName in typeNodeJson) {
      const { schemaPath } = typeNodeJson[typeName];
      if (schemaPath) {
        pathTypeMap.set(schemaPath, typeName);
      }
    }
  }
  const cwd = getCwd(ramlAbsolutePath);
  for (const resourceNode of ramlAPI.allResources()) {
    for (const methodNode of resourceNode.methods()) {
      let requestSchema: string | JSONSchemaObject;
      let responseSchema: string;
      const method = methodNode.method().toUpperCase() as HTTPMethod;
      const fieldName = methodNode.displayName();
      const description = methodNode.description()?.value() || resourceNode.description()?.value();
      let fullRelativeUrl = resourceNode.completeRelativeUri();
      for (const uriParameterNode of resourceNode.uriParameters()) {
        const paramName = uriParameterNode.name();
        fullRelativeUrl = fullRelativeUrl.replace(`{${paramName}}`, `{args.${paramName}}`);
      }
      for (const queryParameterNode of methodNode.queryParameters()) {
        requestSchema = requestSchema || {
          type: 'object',
          properties: {},
          required: [],
        };
        const parameterName = queryParameterNode.name();
        const queryParameterNodeJson = queryParameterNode.toJSON();
        if (queryParameterNodeJson.required) {
          (requestSchema as JSONSchemaObject).required.push(parameterName);
        }
        if ('enum' in queryParameterNodeJson) {
          (requestSchema as JSONSchemaObject).properties[parameterName] = {
            type: 'string',
            enum: queryParameterNodeJson.enum,
          };
        } else {
          (requestSchema as JSONSchemaObject).properties[parameterName] = toJsonSchema(queryParameterNodeJson.example, {
            required: false,
            strings: {
              detectFormat: true,
            },
          });
        }
        if (queryParameterNode.description) {
          (requestSchema as JSONSchemaObject).properties[parameterName].description =
            queryParameterNodeJson.description;
        }
      }

      for (const bodyNode of methodNode.body()) {
        if (bodyNode.name().includes('application/json')) {
          requestSchema = bodyNode.toJSON().schemaPath;
        }
      }
      for (const responseNode of methodNode.responses()) {
        if (responseNode.code().value().startsWith('2')) {
          for (const bodyNode of responseNode.body()) {
            if (bodyNode.name().includes('application/json')) {
              responseSchema = bodyNode.toJSON().schemaPath;
            }
          }
        }
      }
      if (fieldName) {
        const operationType: any = method === 'GET' ? 'query' : 'mutation';
        const graphQLFieldName = sanitizeNameForGraphQL(fieldName);
        operations.push({
          type: operationType,
          field: graphQLFieldName,
          description,
          path: fullRelativeUrl,
          method,
          requestSchema,
          requestTypeName: typeof requestSchema === 'string' ? pathTypeMap.get(requestSchema) : undefined,
          responseSchema,
          responseTypeName: pathTypeMap.get(responseSchema),
        });
      }
    }
  }
  return {
    operations,
    baseUrl,
    cwd,
    fetch,
  };
}
