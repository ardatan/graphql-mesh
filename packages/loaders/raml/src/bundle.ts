import {
  getGraphQLSchemaFromBundle,
  createBundle as createJSONSchemaLoaderBundle,
  JSONSchemaLoaderBundle as RAMLLoaderBundle,
} from '@omnigraph/json-schema';
import { getJSONSchemaOptionsFromRAMLOptions } from './getJSONSchemaOptionsFromRAMLOptions';
import { RAMLLoaderOptions } from './types';

export async function createBundle(name: string, ramlLoaderOptions: RAMLLoaderOptions): Promise<RAMLLoaderBundle> {
  const { operations, baseUrl, cwd, fetch } = await getJSONSchemaOptionsFromRAMLOptions(ramlLoaderOptions);
  return createJSONSchemaLoaderBundle(name, { ...ramlLoaderOptions, baseUrl, operations, cwd, fetch });
}

export { getGraphQLSchemaFromBundle, RAMLLoaderBundle };
