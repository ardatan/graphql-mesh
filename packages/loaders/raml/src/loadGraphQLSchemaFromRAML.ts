import { loadGraphQLSchemaFromJSONSchemas } from '@omnigraph/json-schema';
import { getJSONSchemaOptionsFromRAMLOptions } from './getJSONSchemaOptionsFromRAMLOptions';
import { RAMLLoaderOptions } from './types';

/**
 * Creates a local GraphQLSchema instance from a RAML API Document.
 * Everytime this function is called, the RAML file and its dependencies will be resolved on runtime.
 * If you want to avoid this, use `createBundle` function to create a bundle once and save it to a storage
 * then load it with `loadGraphQLSchemaFromBundle`.
 */
export async function loadGraphQLSchemaFromRAML(name: string, options: RAMLLoaderOptions) {
  const extraJSONSchemaOptions = await getJSONSchemaOptionsFromRAMLOptions(options);
  return loadGraphQLSchemaFromJSONSchemas(name, {
    ...options,
    ...extraJSONSchemaOptions,
  });
}
