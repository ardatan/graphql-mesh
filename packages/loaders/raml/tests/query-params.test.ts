import { execute, GraphQLSchema, parse } from 'graphql';
import { printSchemaWithDirectives } from '@graphql-tools/utils';
import { Response, URL } from '@whatwg-node/fetch';
import { loadGraphQLSchemaFromRAML } from '../src/loadGraphQLSchemaFromRAML';

describe('Query Parameters', () => {
  let schema: GraphQLSchema;
  beforeAll(async () => {
    schema = await loadGraphQLSchemaFromRAML('query-params', {
      source: './fixtures/query-params.raml',
      cwd: __dirname,
      async fetch(url: string) {
        const parsedUrl = new URL(url);
        if (parsedUrl.searchParams.get('isApproved') === 'true') {
          return Response.json([
            {
              id: '1',
              name: 'Approved Product',
            },
          ]);
        } else {
          return Response.json([
            {
              id: '2',
              name: 'Not Approved Product',
            },
          ]);
        }
      },
    });
  });
  it('generates correct schema', () => {
    expect(printSchemaWithDirectives(schema)).toMatchSnapshot();
  });
  it('respects default values', async () => {
    const result = await execute({
      schema,
      document: parse(/* GraphQL */ `
        query {
          listOfProducts {
            id
            name
          }
        }
      `),
    });
    expect(result.errors).toBeFalsy();
    expect(result.data).toEqual({
      listOfProducts: [
        {
          id: '1',
          name: 'Approved Product',
        },
      ],
    });
  });
  it('respects arguments with values', async () => {
    const result = await execute({
      schema,
      document: parse(/* GraphQL */ `
        query {
          falseOne: listOfProducts(isApproved: false) {
            id
            name
          }
          trueOne: listOfProducts(isApproved: true) {
            id
            name
          }
        }
      `),
    });
    expect(result.errors).toBeFalsy();
    expect(result.data).toEqual({
      trueOne: [
        {
          id: '1',
          name: 'Approved Product',
        },
      ],
      falseOne: [
        {
          id: '2',
          name: 'Not Approved Product',
        },
      ],
    });
  });
});
