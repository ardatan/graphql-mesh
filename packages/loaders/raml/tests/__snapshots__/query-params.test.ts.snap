// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query Paramters generates correct schema 1`] = `
"schema {
  query: Query
}

directive @globalOptions(sourceName: String, endpoint: String, operationHeaders: ObjMap, queryStringOptions: ObjMap, queryParams: ObjMap) on OBJECT

directive @httpOperation(path: String, operationSpecificHeaders: ObjMap, httpMethod: HTTPMethod, isBinary: Boolean, requestBaseBody: ObjMap, queryParamArgMap: ObjMap, queryStringOptionsByParam: ObjMap) on FIELD_DEFINITION

type Query @globalOptions(sourceName: "query-params", endpoint: "http://localhost:3000") {
  "Get a list of products"
  listOfProducts(
    "Get approved products"
    isApproved: Boolean = true
  ): [Product] @httpOperation(path: "/listOfProducts", httpMethod: GET, queryParamArgMap: "{\\"isApproved\\":\\"isApproved\\"}")
}

type Product {
  id: String
  name: String
}

scalar ObjMap

enum HTTPMethod {
  GET
  HEAD
  POST
  PUT
  DELETE
  CONNECT
  OPTIONS
  TRACE
  PATCH
}"
`;
